### PR TITLE
Add wayland primary selection clipboard support

### LINKED ISSUE #2510: Support primary selection clipboard on Wayland
Currently the native Wayland clipboard only supports the 'normal' clipboard, however it does not support the very commonly used primary selection clipboard, which stores the last selection made.

This requires making some changes to the smithay-clipboard crate first. @trimental do you know how much work this would be approximately? I'm not sure what exactly needs to be done to extend the existing code to also support this.

### PR DESCRIPTION
**Pull request by chrisduerr**
This is a follow-up PR to #2513.

Using the smithay-clipboard v0.3 release, this adds support for the primary selection clipboard on Wayland.

Since the progress on the rust-clipboard PR has been a bit slow, the clipboard crate has also been moved into alacritty under the name copypasta (same as the previous clipboard crate we had). This should hopefully allow directly sending PRs, without having it live in a separate repo.

Fixes #2510.

### CODE REVIEW

**Comment by chrisduerr** [comment]
@trimental I think @cole-h has mentioned that he had troubles getting Alacritty to work with copy/pasting between Alacritty and Firefox using the selection clipboard, are you aware of any issues like that?

I'm not sure what exact combinations he's tried, though I'd recommend also trying copy/paste between two Firefox windows and Alacritty with other terminal emulators (I think gnome-terminal might be wayland-native?, or Kitty?).

**Comment by cole-h** [comment]
I did indeed. I likely won't be able to get to test gnome-terminal or Kitty today (sorry!), but I'll try to get around to doing some testing and documenting my STR sometime tomorrow. I might even whip out OBS and get some screen recordings.

**Comment by chrisduerr** [comment]
@cole-h No worries, all help is very appreciated! If I'm in a hurry I can always test things myself. :)

**Comment by trimental** [comment]
@chrisduerr @cole-h Ah yes I forgot to advertise the offering type. Should be fixed in 0.3.1

**Comment by cole-h** [comment]
I can confirm bumping `smithay-clipboard` to `0.3.1` lets me middle-click from Alacritty to Firefox (and vice versa, but that had always worked). Glad it was a simple fix, thank you!

File: copypasta/README.md [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+
+```rust
+extern crate copypasta;
+
+use copypasta::ClipboardContext;
+
+fn example() {
+    let mut ctx = ClipboardContext::new().unwrap();
+    println!("{:?}", ctx.get_contents());
+    ctx.set_contents("some string".to_owned()).unwrap();
```
Comment (nixpulvis) [inline copypasta/README.md]: Maybe show a read here too.

File: copypasta/examples/hello_world.rs [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+
+use copypasta::ClipboardContext;
+use copypasta::ClipboardProvider;
+
+fn main() {
+    let mut ctx = ClipboardContext::new().unwrap();
+
+    let the_string = "Hello, world!";
+
+    ctx.set_contents(the_string.to_owned()).unwrap();
```
Comment (nixpulvis) [inline copypasta/examples/hello_world.rs]: I'd personally add the following to both examples:

```rust
println!("Copied contents to clipboard until exit...");
loop {}
```

File: copypasta/README.md (hunk 1 above)
Comment (chrisduerr) [reply copypasta/README.md]: I'd like to keep the copypasta crate mostly unchanged for now. It has been months now without any response, but I still haven't completely given up on upstreaming our changes.

So at least for this release, I think it's better to mostly keep it untouched. If we decide to go over it again and modify/clean some stuff up, there's probably a few things that could be improved.

Comment (nixpulvis) [reply copypasta/README.md]: Oh that's fair.

**Pull request merged**